Frame lowering optimizations

nikolaperic opened the pull request:
Enable save/restore generation when there is a gap between registers by inserting the missing ones at the moment when spill slots for callee-saved registers are assigned.

Commit:
Enable save/restore generation when there is a gap between registers

Enable save/restore generation when there is a gap between registers
by inserting the missing ones at the moment when spill slots for
callee-saved registers are assigned.

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -935,21 +935,57 @@ bool MipsSEFrameLowering::assignCalleeSavedSpillSlots(
+
+  auto MaxCS = std::max_element(
+      CSI.begin(), CSI.end(), [&Regs](CalleeSavedInfo &a, CalleeSavedInfo &b) {
+        return Regs.at(a.getReg()) < Regs.at(b.getReg());
```
This (and at line 966-968) is really the same comparison function as `SortCalleeSaves`; can we rename that to `CompareCalleeSaves` and re-use here and in 966?

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -935,21 +935,57 @@ bool MipsSEFrameLowering::assignCalleeSavedSpillSlots(
   if (!STI.hasNanoMips())
     return false;
 
+  std::unordered_map<unsigned, unsigned> Regs{
```
While were at it, this map is constant, so can we add it as a static const member?

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -935,21 +935,57 @@ bool MipsSEFrameLowering::assignCalleeSavedSpillSlots(
+  for (unsigned i = MinCSNum + 1; i < MaxCSNum; i++) {
+
+    auto SearchResult =
+        std::find_if(CSI.begin(), CSI.end(), [&Regs, &i](CalleeSavedInfo &p) {
```
It seems a bit inefficient to do a linear find for each potential register.

How about, instead:
  - Go through CSI and build a BitSet of CSNums that are in the saves list
  - Iterate over CSNums, any CSNum not in BitSet must be added
  - Find regnum for CSI by `find_if` (this is linear time complexity, but is rare), OR build an inverse mapping `unordered_set` (again, static member) that maps CSNum to register numbe , and use that.

?

nikolaperic commented:
Thanks for the suggestions

Commit:
NanoMips: Two-step stack setup

Adjust stack pointer in two steps if offset is larger than 4096.
In the first step adjust stack pointer for the size necessary to
spill CSR onto the stack. In the second step adjust stack pointer
for the size necessary to spill local objects.

Commit:
NanoMips: Frame pointer setup

If function uses FP, it will now point to address -4096 from the
beginning of function's stack. After FP setup following offsets
will be relative to SP if function has no var-sized objects. If it
has var-sized objects offsets will be relative to FP.
Also, stack realignment now happens with INS instruction.

nikolaperic commented:
Added two new commits that address other issues, but that rely on changes from the previous save/restore optimization commit.

In the first newly added commit, two-step SP setup is implemented if offset is greater than 4096. This fixes FrameLowering for large offsets. 

In the second newly added commit, the FP setup is implemented so that if function wants to use it, FP will point to an address at -4096 from the beginning of function's stack.

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -432,19 +432,61 @@ void MipsSEFrameLowering::emitPrologue(MachineFunction &MF,
+  const std::vector<CalleeSavedInfo> &CSI = MFI.getCalleeSavedInfo();
+
+  int64_t CalleeSavedStackSize;
+  int64_t LoaclStackSize;
```
"*Local*StackSize"?

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -432,19 +432,61 @@ void MipsSEFrameLowering::emitPrologue(MachineFunction &MF,
+
+    // Move MBBI_2 to point to the first instruction after
+    // calle-saves store sequence. That's the place for the second
+    // steck pointer adjustment.
```
steck -> stack

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -432,19 +432,61 @@ void MipsSEFrameLowering::emitPrologue(MachineFunction &MF,
+    // Move MBBI_2 to point to the first instruction after
+    // calle-saves store sequence. That's the place for the second
+    // steck pointer adjustment.
+    std::advance(MBBI_2, NumOfCSI);
```
This is assuming that the prologue *definitely* has one instruction per callee save, and that they're right here? That might be the case but in future where save/restore are generated might invalidate this. It would be a good idea to at least `assert` that the instructions are what we assume them to be and document that assumption.

nikolaperic replied:
Yes, I think that's the safer way. However, in order to do that I think we need to introduce a new function which checks opcodes and operands of skipped instructions. This function will be nanomips-specific  so I think it's not a good idea to write it inside ```MipsSEFrameLowering```.  Maybe a good place to write such function is inside ```MipsFunctionInfo``` since we have already put ```CalleeSavedStackSize``` variable there? Is this a good approach for this problem?  

Note that on the line [452](https://github.com/MediaTek-Labs/llvm-project/pull/1/files#diff-2436c22f93c462e0e51bcd9e774e00bfcd69bac4b0246db61fe28a1f796b8397L452) we already had this kind of skipping CSR store instructions.

cme commented on llvm/lib/Target/Mips/MipsSEFrameLowering.cpp:
```diff
@@ -432,19 +432,61 @@ void MipsSEFrameLowering::emitPrologue(MachineFunction &MF,
+    // Adjust stack.
+    TII.adjustStackPtr(SP, -StackSize, MBB, MBBI);
+
+  if (MipsFI->isTwoStepStackSetup(MF)) {
```
Any reason to keep these as separate 'if' conditionals on the same condition `MipsFI->isTwoStepStackSetup`? Could be merged with the above?

nikolaperic replied:
Yes, you are right, there is no reason to keep them separated. I will merge them. Thanks!

djtodoro commented:
@nikolaperic Please set the target branch to ```nanomips-llvm13-submission```.

Commit:
Merged same if instructions

djtodoro commented:
@nikolaperic I think we picked different approach for this, right? Can we close this (and make sure we do have it in LLVM 16)?
